src/main/java/org/primeframework/mvc/security/BaseUserIdCookieSecurityContext.java

```diff
@@ -0,0 +1,257 @@
+ */
+package org.primeframework.mvc.security;
+
+import javax.crypto.BadPaddingException;
+import java.time.Clock;
+import java.time.Duration;
+import java.time.ZonedDateTime;
+import java.util.Optional;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.fusionauth.http.Cookie;
+import io.fusionauth.http.server.HTTPRequest;
+import io.fusionauth.http.server.HTTPResponse;
+import org.primeframework.mvc.ErrorException;
+import org.primeframework.mvc.util.CookieTools;
+
+/**
+ * Logs in and out users based on storing their User ID in a cookie and also adds a sliding session
+ * timeout with a maximum age
+ *
+ * @author Brady Wied
+ */
+public abstract class BaseUserIdCookieSecurityContext<TUserId> implements UserLoginSecurityContext {
+  public static final String UserKey = "primeCurrentUser";
+
+  private static final String ContextKey = "primeLoginContext";
+
+  protected final CookieProxy sessionCookie;
+
+  private final Clock clock;
+
+  private final Encryptor encryptor;
+
+  private final ObjectMapper objectMapper;
+
+  private final HTTPRequest request;
+
+  private final HTTPResponse response;
+
+  private final Duration sessionMaxAge;
+
+  private final Duration sessionTimeout;
+
+  protected BaseUserIdCookieSecurityContext(HTTPRequest request, HTTPResponse response, Encryptor encryptor, ObjectMapper objectMapper,
+                                            Clock clock, Duration sessionTimeout, Duration sessionMaxAge) {
+    this.request = request;
+    this.response = response;
+    this.encryptor = encryptor;
+    this.objectMapper = objectMapper;
+    this.clock = clock;
+    this.sessionMaxAge = sessionMaxAge;
+    this.sessionTimeout = sessionTimeout;
+    long timeoutInSeconds = sessionTimeout.toSeconds();
+    this.sessionCookie = new CookieProxy(getCookieName(), timeoutInSeconds, Cookie.SameSite.Strict);
+  }
+
+  /**
+   * Get the currently logged in user. Only calls retrieveUserById once per request cycle.
+   *
+   * @return null if no user is logged in. Otherwise the user object as retrieved by retrieveUserById using the ID from the session
+   */
+  @Override
+  public Object getCurrentUser() {
+    var user = request.getAttribute(UserKey);
+    if (user != null) {
+      return user;
+    }
+    UserIdSessionContext<TUserId> sessionContext = resolveContext();
+    if (sessionContext == null) {
+      return null;
+    }
+    user = retrieveUserById(sessionContext.getUserId());
+    this.request.setAttribute(UserKey, user);
+    return user;
+  }
+
+  /**
+   * The current session ID
+   *
+   * @return the session ID, if a session exists, otherwise null
+   */
+  @Override
+  public String getSessionId() {
+    return Optional.ofNullable(resolveContext())
+                   .map(UserIdSessionContext::getSessionId)
+                   .orElse(null);
+  }
+
+  @Override
+  public boolean isLoggedIn() {
+    // user MIGHT a DB query to hydrate the user so we can do a lighter weight check
```

Comment:
Not sure what this comment means, can it be re-worded for clarity?

Reply:
Fixed 2ffc886bd9183800a4702086bf860df83ced6c90